// htmlnano/lib/modules/mergeScripts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = mergeScripts;

/* Merge multiple <script> into one */
function mergeScripts(tree) {
  var scriptNodesIndex = {};
  var scriptSrcIndex = 1;
  tree.match({
    tag: 'script'
  }, function (node) {
    var nodeAttrs = node.attrs || {};

    if (nodeAttrs.src) {
      scriptSrcIndex++;
      return node;
    }

    var scriptType = nodeAttrs.type || 'text/javascript';

    if (scriptType !== 'text/javascript' && scriptType !== 'application/javascript') {
      return node;
    }

    var scriptKey = JSON.stringify({
      id: nodeAttrs.id,
      "class": nodeAttrs["class"],
      type: scriptType,
      defer: nodeAttrs.defer !== undefined,
      async: nodeAttrs.async !== undefined,
      index: scriptSrcIndex
    });

    if (!scriptNodesIndex[scriptKey]) {
      scriptNodesIndex[scriptKey] = [];
    }

    scriptNodesIndex[scriptKey].push(node);
    return node;
  });

  var _loop = function _loop() {
    var scriptNodes = _Object$values[_i];
    var lastScriptNode = scriptNodes.pop();
    scriptNodes.reverse().forEach(function (scriptNode) {
      var scriptContent = (scriptNode.content || []).join(' ');
      scriptContent = scriptContent.trim();

      if (scriptContent.slice(-1) !== ';') {
        scriptContent += ';';
      }

      lastScriptNode.content = lastScriptNode.content || [];
      lastScriptNode.content.unshift(scriptContent);
      scriptNode.tag = false;
      scriptNode.content = [];
    });
  };

  for (var _i = 0, _Object$values = Object.values(scriptNodesIndex); _i < _Object$values.length; _i++) {
    _loop();
  }

  return tree;
}
